### ISSUE TITLE
[usrloc] Event E_UL_CONTACT_INSERT works incorrectly in db_mode 3

### ISSUE THREAD

**Issue by ivanoff946** [OpenSIPS/opensips#1573]
Hello collegues! I use opensips proxy for work with similarly scenario: https://blog.opensips.org/2017/03/10/advanced-sip-scenarios-with-event-based-routing/
but i have some errors.

**Code:**
```
route {
...
        t_newtran();
        t_wait_for_new_branches();
        $avp(filter) = "aor=" + $tU + "@" + $td;
        notify_on_event("E_UL_CONTACT_INSERT","$avp(filter)","INSERT_CT","120");
        route(RELAY);
...
}

route[INSERT_CT] {
    xlog("L_INFO", "[INSERT_CT] New contact [$avp(uri)] registered for user [$avp(aor)], inserting new branch to ongoing call\r\n");
    usleep("1000");
    t_inject_branches("event");
}
```
**Scenario:** opensips receives INVITE to AOR from softswitch. Opensips waits new successfull REGISTER requests and forks call to new Contacts for this AOR. 

From usrloc module doc: E_UL_CONTACT_INSERT - This event is raised **when a new contact is inserted** in any of the existing AOR's contact list. For each new contact, if its AOR does not exist in the memory, then both the E_UL_AOR_CREATE and E_UL_CONTACT_INSERT events will be raised.

In real life i have this situation. One AOR can use on several softphones. 
Step by step, problem scenario:
1) Usrloc table BEFORE call to AOR (two softphones):
```
sip=# select username,domain,contact,received from location2 where username = 'user3368';
 username  |   domain    |               contact               |        received        |       user_agent
-----------+-------------+-------------------------------------+------------------------+-------------------------
 user3368 | domain.com | sip:user3368@1.1.1.110:26519;ob | sip:1.1.1.110:26519 | A
 user3368 | domain.com | sip:user3368@1.1.1.110:59426;ob | sip:1.1.1.110:59426 | B
```
2) opensips receives INVITE to AOR from softswitch and generate push requests by two softphones;
3) Two softphones received push request and generate REGISTER requests to opensips;
4) First, opensips receives REGISTER with auth_data from softfone with user_agent **A**, in logs i see:
```
Dec 18 13:47:46 opensips-pres /usr/sbin/opensips[30793]: [MAIN] Incoming request (M=INVITE IP=udp:1.1.1.110:2838
Dec 18 13:47:47 sip /usr/sbin/opensips[30793]: [INSERT_CT] New contact [sip:user3368@1.1.1.110:59426;ob] registered for user [user3368@domain.com], inserting new branch to ongoing call
Dec 18 13:47:47 sip /usr/sbin/opensips[30793]: [INSERT_CT] New contact [sip:user3368@1.1.1.110:2838;ob] registered for user [user3368@domain.com], inserting new branch to ongoing call
```
5) Second, opensips receives REGISTER with auth_data from softfone with user_agent **B**, in logs i see:
```
Dec 18 13:47:46 opensips-pres /usr/sbin/opensips[30793]: [MAIN] Incoming request (M=INVITE IP=udp:1.1.1.110:47564
Dec 18 13:47:47 sip /usr/sbin/opensips[30793]: [INSERT_CT] New contact [sip:user3368@1.1.1.110:2838;ob] registered for user [user3368@domain.com], inserting new branch to ongoing call
Dec 18 13:47:47 sip /usr/sbin/opensips[30793]: [INSERT_CT] New contact [sip:user3368@1.1.1.110:47564;ob] registered for user [user3368@domain.com], inserting new branch to ongoing call
```
Here, event E_UL_CONTACT_INSERT raised **for every already existing registration**.
In logs you may see, that for **one** REGISTER E_UL_CONTACT_INSERT raised **twice (because we have two entries in usrloc for AOR user3368@domain.com)**.

In other words, E_UL_CONTACT_INSERT raised N*N times, but not N times (N - entries in usrloc).
Injecting branch **must be for only one Contact** from successfull REGISTER. Now i see that injecting branch exec for every entries in usrloc, and so for everyone new successfull REGISTER.

opensips -V:
```
version: opensips 2.4.3 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, F_MALLOC, FAST_LOCK-FUTEX-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: 5f1ec37
main.c compiled on 16:32:14 Dec  4 2018 with gcc 4.8.5
```
Thanks!

**Comment by bogdan-iancu** [comment]
Please provide the following input:
1) the output of `opensipsctl ul show` after opensips started and before making the test call
2) opensips logs in level 4 covering (a) the initial call, (b) the triggered registrations (due PN) and (c) the new resulting branches.
A short listing of the parties (as SIP aors) involved in the call will help.

In case of sensitive data, please provide the data / link off-ticket.

**Comment by ivanoff946** [comment]
Hello @bogdan-iancu !

1) the output of opensipsctl ul show after opensips started and before making the test call
I use this settings for module "usrloc.so":
```
modparam("usrloc", "db_mode", 3)
```
Option db_mode=3 because we have situations when softphones reregister very often times. Every new call to SoftPhone(SP) generates push to SP, and SP generate new REGISTER.
One SIP account may be registered on some SP. Every SP have hie individual ID(you can see it in user_agent column below).
I can manage amount of registrations use this code:
```
route {
...
    if (is_method("REGISTER")) route(REGISTER);
...

route[REGISTER] {
...
    avp_db_query("DELETE FROM location2 WHERE username = '$tU' AND domain = '$td' AND user_agent = '$ua'");
...
    if (!save("location2", "e60"))
    {
        xlog("L_ERR", "[REGISTER] Saving contact failed LOGHDR");
        sl_reply_error();
        exit;
    }
```

The meaning of problem, i think, it is "select" SQL request in save() procedure. What i mean - for example in table location i have two registrations before call:
```
opensips=# select * from location2;
     contact_id      |  username   |      domain       |                contact                |        received        | path |  expires   | q |                callid                | cseq  |    last_modif
ied    | flags | cflags |                user_agent                 |       socket        | methods | sip_instance | kv_store | attr 
---------------------+-------------+-------------------+---------------------------------------+------------------------+------+------------+---+--------------------------------------+-------+--------------
-------+-------+--------+-------------------------------------------+---------------------+---------+--------------+----------+------
 2595832603969860188 | user1 | example.domain.com | sip:user1@1.1.1.110:51559;ob | sip:1.1.1.110:51559 |      | 1546263281 | 0 | caff18cb-db90-486a-bd19-020cca76fcb4 | 54046 | 2018-12-24 16
:34:41 |     0 |        | CALLER 83a1b923f5ea809c3e45d722f9e9f6f  | udp:1.1.6.9:5060 |    8063 |              |          | 
                  49 | user1 | example.domain.com | sip:user1@1.1.1.110:60919;ob | sip:1.1.1.110:60919 |      | 1546413469 | 0 | c2be39fa-7dde-4aba-ba67-bc488b68651c | 18847 | 2018-12-26 10
:17:49 |     0 |        | CALLER a26123f684cbfea5571d64b485f580ff | udp:1.1.6.9:5060 |    8063 |              |          | 
(2 rows)
```

When i call to SIP account user1@example.domain.com, push request sends to two SoftPhones(SP) with id "83a1b923f5ea809c3e45d722f9e9f6f" and "a26123f684cbfea5571d64b485f580ff".
For example, SP id "a26123f684cbfea5571d64b485f580ff" woke up first and generate register. In config above wee see that opensips try to delete existing registration for SP id "a26123f684cbfea5571d64b485f580ff" before do save() operation.
In level 4 logs i see this:
```
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: [MAIN] Incoming request (M=REGISTER IP=udp:1.1.1.110:58842 fromIP=udp:1.1.1.110:58842 RURI=sip:example.domain.com
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:avpops:ops_dbquery_avps: query [DELETE FROM location2 WHERE username = 'user1' AND domain = 'example.domain.com' AND user_agent = 'M.TAL
KER a26123f684cbfea5571d64b485f580ff']
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:db_postgres:db_postgres_submit_query: 0x7f8fe59707c8 PQsendQuery(DELETE FROM location2 WHERE username = 'user1' AND domain = 'example.domain.com' AND user_agent = 'CALLER a26123f684cbfea5571d64b485f580ff')
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:core:db_new_result: allocate 48 bytes for result set at 0x7f8fe5419260
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:db_postgres:db_postgres_store_result: 0x7f8fe59707c8 PQresultStatus(PGRES_COMMAND_OK) PQgetResult(0xb40250)
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:db_postgres:free_query: PQclear(0xb40250) result set
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:avpops:db_query_avp: no result after query
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:avpops:db_close_query: close avp query
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:db_postgres:db_postgres_submit_query: 0x7f8fe5972ca0 PQsendQuery(select contact_id,contact,expires,q,callid,cseq,flags,cflags,user_agent,received,path,socket,methods,last_modified,sip_instance,kv_store,attr from location2 where username='user1' AND domain='example.domain.com' order by q)
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:core:db_new_result: allocate 48 bytes for result set at 0x7f8fe5419260
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:db_postgres:db_postgres_store_result: 0x7f8fe5972ca0 PQresultStatus(PGRES_TUPLES_OK) PQgetResult(0xb40250)
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:db_postgres:db_postgres_get_columns: 17 columns returned from the query
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:db_postgres:db_postgres_convert_rows: PQgetvalue(0x7f8fe5972ca0,0,1)=[sip:user1@1.1.1.110:51559;ob]
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:db_postgres:db_postgres_convert_rows: allocated 37 bytes for row_buf[1] at 0x7f8fe5c69d50
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:db_postgres:db_postgres_convert_rows: [0][1] Column[contact]=[sip:user1@1.1.1.110:51559;ob]
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:core:evi_raise_event_msg: found subscriber E_UL_CONTACT_INSERT
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:event_routing:notify_ebr_subscriptions: notification received for event E_UL_CONTACT_INSERT, checking subscriptions
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:event_routing:notify_ebr_subscriptions: key <aor> found, checking value
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:event_routing:notify_ebr_subscriptions: subscription type [NOTIFY]from process 24(pid 5146) matched event, generating notification via IPC
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:core:evi_raise_event_msg: found subscriber E_UL_CONTACT_INSERT
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:event_routing:notify_ebr_subscriptions: notification received for event E_UL_CONTACT_INSERT, checking subscriptions
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:event_routing:notify_ebr_subscriptions: key <aor> found, checking value
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:event_routing:notify_ebr_subscriptions: subscription type [NOTIFY]from process 24(pid 5146) matched event, generating notification via IPC
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:registrar:build_contact: created Contact HF: Contact: <sip:user1@1.1.1.110:51559;ob>;q=0;expires=452368;received="sip:1.1.1.110:51559", <sip:user1@1.1.1.110:58842;ob>;q=0;expires=604800;received="sip:1.1.1.110:58842"
ec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:db_postgres:db_postgres_submit_query: 0x7f8fe5972ca0 PQsendQuery(insert into location2 (username,contact,expires,q,callid,cseq,flags,cflags,user_agent,received,path,socket,methods,last_modified,sip_instance,kv_store,attr,domain ) values ('user1','sip:user1@1.1.1.110:58842;ob'1546415713,0.00      ,'8994d05d-1e92-4f79-8df2-7e2df8a919dd',63322,0,'','CALLER a26123f684cbfea5571d64b485f580ff','sip:1.1.1.110:58842',NULL,'udp:1.1.6.9:5060',8063,'2018-12-26 10:55:13',NULL,NULL,NULL,'example.domain.com'))
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: DBG:core:db_new_result: allocate 48 bytes for result set at 0x7f8fe5419260
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: [REGISTER] Registered successfully (M=REGISTER IP=udp:1.1.1.110:58842 fromIP=udp:1.1.1.110:58842 RURI=sip:example.domain.com DURI=<null> F=sip:user1@example.domain.com T=sip:user1@example.domain.com oP=udp dP=<null> rP=udp ID=8994d05d-1e92-4f79-8df2-7e2df8a919dd ct=<sip:user1@1.1.1.110:58842;ob> cseq=63322 UA=CALLER a26123f684cbfea5571d64b485f580ff)
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: [INSERT_CT] New contact [sip:user1@1.1.1.110:51559;ob] registered for user [user1@example.domain.com], inserting new branch to ongoing call
Dec 26 10:55:13 opensips /usr/sbin/opensips[5149]: [INSERT_CT] New contact [sip:user1@1.1.1.110:58842;ob] registered for user [user1@example.domain.com], inserting new branch to ongoing call
```
In logs:
1) Income new REGISTER for SP id "a26123f684cbfea5571d64b485f580ff"
2) Delete (if existing) registration for SP id "a26123f684cbfea5571d64b485f580ff" in DB
3) Select from location2 for user1@example.domain.com. Here opensips get remaining one registration for SP id "83a1b923f5ea809c3e45d722f9e9f6f"
4) E_UL_CONTACT_INSERT raised twice
5) [INSERT_CT] New contact [sip:user1@1.1.1.110:51559;ob] - for remaining registration for SP id "83a1b923f5ea809c3e45d722f9e9f6f" <- **Any chance to fix it?**
6) [INSERT_CT] New contact [sip:user1@1.1.1.110:58842;ob] - for new incoming REGISTER request for "a26123f684cbfea5571d64b485f580ff"

**Comment by bogdan-iancu** [comment]
Hey @ivanoff946, I managed to isolate the problem - it seems related to the db_mode 3 in usrloc. When db only is used, the CONTACT_INSERT event is triggered for *all* the contacts belonging to the AOR under processing - this because the contacts are first loaded from DB and temporary pushed into memory cash (here the event is triggered).
Switch to db_mode 2 and let's see if the problem still replicate. BTW, 3 is the slowest one when comes to performance.

**Comment by ivanoff946** [comment]
Hello @bogdan-iancu ! I was trying works with 'db_mode = 2', but it not help. In 'db_mode = 3' i can operate entries in usrloc: 
```
avp_db_query("DELETE FROM location2 WHERE username = '$tU' AND domain = '$td' AND user_agent = '$ua'");  <<<<----- I mean this instruction
...
    if (!save("location2", "e60"))
    {
        xlog("L_ERR", "[REGISTER] Saving contact failed LOGHDR");
        sl_reply_error();
        exit;
    }
```
Softphones always create new registrations (mobile phones), so i want to store in usrloc only **one** entry for a specific softphone instance (all softphones have unic ID in user-agent).

If i use 'db_mode = 2', instruction "avp_db_query("DELETE FROM location2..." delete entry in DB, but not delete entry in usrloc memory. So, one softphone for in a short time may create a lot of registrations.
Maybe i can operate entries in usrloc memory and delete them by user-agent matching?

**Comment by bogdan-iancu** [comment]
@ivanoff946 , why do you want to do manual ops on the `location` table ? you shouldn;t do that and let OpenSIPS to updated its table. In `db_mode` 2, the table is updated by opensips to reflect the contacts it has in memory - Note that the primary storage is memory, the date is only flush to DB (never read during runtime).
So, the key question is : in `db_mode` 2, is the CONTACT_INSERT event properly triggered ?
